File: csharp/AppEncryption/AppEncryption.Tests/AppEncryption/Persistence/AdhocPersistenceTest.cs
using System;
using GoDaddy.Asherah.AppEncryption.Persistence;
using LanguageExt;
using Moq;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.AppEncryption.Persistence
{
    [Collection("Logger Fixture collection")]
    public class AdhocPersistenceTest
    {
        private readonly Mock<Func<string, Option<string>>> persistenceLoadMock;
        private readonly Mock<Action<string, string>> persistenceStoreMock;

        private readonly Persistence<string> adhocPersistence;

        public AdhocPersistenceTest()
        {
            persistenceLoadMock = new Mock<Func<string, Option<string>>>();
            persistenceStoreMock = new Mock<Action<string, string>>();
            adhocPersistence = new AdhocPersistence<string>(persistenceLoadMock.Object, persistenceStoreMock.Object);
        }

        [Fact]
        private void TestConstructor()
        {
            Persistence<string> adhocPersistence = new AdhocPersistence<string>(persistenceLoadMock.Object, persistenceStoreMock.Object);
            Assert.NotNull(adhocPersistence);
        }

        [Fact]
        private void TestLoad()
        {
            const string expectedValue = "some_value";
            const string someKey = "some_key";
            persistenceLoadMock.Setup(x => x(It.IsAny<string>())).Returns(expectedValue);
            Option<string> actualValue = adhocPersistence.Load(someKey);

            Assert.Equal(expectedValue, actualValue);
            persistenceLoadMock.Verify(x => x(someKey));
        }

        [Fact]
        private void TestStore()
        {
            const string key = "some_key";
            const string value = "some_value";

            adhocPersistence.Store(key, value);

            persistenceStoreMock.Verify(x => x(key, value));
        }
    }
}

File: java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/persistence/AdhocPersistenceTest.java
package com.godaddy.asherah.appencryption.persistence;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import java.util.Optional;
import java.util.function.BiConsumer;
import java.util.function.Function;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class AdhocPersistenceTest {
  @Mock
  Function<String, Optional<String>> persistenceLoad;
  @Mock
  BiConsumer<String, String> persistenceStore;
  @InjectMocks
  AdhocPersistence<String> adhocPersistence;

  @Test
  void testConstructor() {
    Persistence<String> adhocPersistence = new AdhocPersistence<>(persistenceLoad, persistenceStore);
    assertNotNull(adhocPersistence);
  }

  @Test
  void testLoad() {
    String expectedValue = "some_value";
    when(persistenceLoad.apply(anyString())).thenReturn(Optional.of(expectedValue));
    
    String key = "some_key";
    Optional<String> actualValue = adhocPersistence.load(key);
    assertTrue(actualValue.isPresent());
    assertEquals(expectedValue, actualValue.get());
    verify(persistenceLoad).apply(key);
  }

  @Test
  void testStore() {
    String key = "some_key";
    String value = "some_value";
    adhocPersistence.store(key, value);
    verify(persistenceStore).accept(key, value);
  }

}
